### PR TITLE
Added Unit Test

### LINKED ISSUE #8: Xamarin : Créer des test unitaires
Objectif : Réaliser des tests unitaires pour l'application xamarin

### LINKED ISSUE #27: Xamarin : Add more UT

### PR DESCRIPTION
**Pull request by Thomas-Julien-SAHUT**

### CODE REVIEW

File: JhipsterXamarin/UnitTestJhispterXamarin/packages.config [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  <package id="Castle.Core" version="4.4.0" targetFramework="net472" />
+  <package id="Moq" version="4.14.7" targetFramework="net472" />
+  <package id="MSTest.TestAdapter" version="2.1.1" targetFramework="net472" />
+  <package id="MSTest.TestFramework" version="2.1.1" targetFramework="net472" />
+  <package id="MvvmCross" version="7.1.2" targetFramework="net472" />
+  <package id="MvvmCross.Tests" version="7.1.2" targetFramework="net472" />
+  <package id="NUnit" version="3.12.0" targetFramework="net472" />
+  <package id="System.Console" version="4.3.1" targetFramework="net472" />
+  <package id="System.IO" version="4.3.0" targetFramework="net472" />
+  <package id="System.Net.Http" version="4.3.4" targetFramework="net472" />
```
Comment (nicolas63) [inline JhipsterXamarin/UnitTestJhispterXamarin/packages.config:13]: To resolve you problem you need to add 
  `<package id="System.Net.Http.Json" version="3.2.1" targetFramework="net472" /> `
here

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Services/AuthenticationService.cs [hunk 2]
```diff
@@ -11,7 +11,7 @@ public class AuthenticationService : IAuthenticationService
         private const string AuthenticationUrl = "api/authenticate";
         private const string AccountUrl = "api/account";
         private const string AuthorizationHeader = "Authorization";
-        private readonly HttpClient _httpClient;
+        public HttpClient _httpClient { get; set; }
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Services/AuthenticationService.cs]: Can you revert this change ? You don't need to expose http client

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Services/IAuthenticationService.cs [hunk 3]
```diff
@@ -7,6 +8,7 @@ public interface IAuthenticationService
     {
         bool IsAuthenticated { get; set; }
         UserModel CurrentUser { get; set; }
+        HttpClient _httpClient { get; set; }
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Services/IAuthenticationService.cs]: Can you revert this change too ?

**Review by nicolas63 — CHANGES_REQUESTED** [review]

File: JhipsterXamarin/UnitTestJhispterXamarin/TestLogin.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            var mockNavLoginService = new Mock<IMvxNavigationService>();
+            loginViewModel = new LoginViewModel(mockNavLoginService.Object, mockAuthService.Object);
+
+            Ioc.RegisterSingleton<IAuthenticationService>(mockAuthService.Object);
+            Ioc.RegisterSingleton<IMvxNavigationService>(mockNavLoginService.Object);
+        }
+        [TestMethod]
+        public void TestSignInAdmin()
+        {
+            initializeAll();
```
Comment (nicolas63) [inline JhipsterXamarin/UnitTestJhispterXamarin/TestLogin.cs]: you must use [TestInitiliaze] or [ClassInitialize] => https://www.automatetheplanet.com/mstest-cheat-sheet/

File: JhipsterXamarin/UnitTestJhispterXamarin/TestLogin.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+            loginModel.Username = "admin";
+            loginModel.Password = "admin";
+            loginModel.RememberMe = false;
+
+            loginViewModel.Username = "admin";
+            loginViewModel.Password = "admin";
+            loginViewModel.RememberMe = false;
+
+            Assert.AreEqual(loginViewModel.GetAuthenticationService().SignIn(loginModel).ToString(), loginViewModel.SignInConnection().ToString(), "Test failed : login error user : admin  | password : admin");
```
Comment (nicolas63) [inline JhipsterXamarin/UnitTestJhispterXamarin/TestLogin.cs]: replace this assertion and use fluent assertion

File: JhipsterXamarin/UnitTestJhispterXamarin/TestService.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+        [Test]
+        public void TestUpdate()
+        {
+            initializeAll();
+            myFirstEntityModelSimple.Name = "the first entity updated";
+            myEntityService.UpdateEntity((MyEntityModel)myFirstEntityModelSimple);
+            var result = myEntityViewModel.CurrentElement;
+            string expected = "the first entity updated";
+            Assert.AreEqual(expected, result.Name, "Test failed: We wanted : " + expected + " We have :  " +result.Name);
```
Comment (nicolas63) [inline JhipsterXamarin/UnitTestJhispterXamarin/TestService.cs]: use fluent assertion

**Review by nicolas63 — CHANGES_REQUESTED** [review]

**Comment by nicolas63** [comment]
this code does not compile ! Can you fix it ?

**Comment by Thomas-Julien-SAHUT** [comment]
I update UT to solve 1 issue on UT but I still have this issue : The Initialization UnitTestJhispterXamarin.TestService.TestViewModel method threw an exception. System.IO.FileNotFoundException: Could not load file or assembly 'System.Net.Http.Json, Version = 3.2.0.0, Culture = neutral, PublicKeyToken = cc7b13ffcd2ddd51' or one of its dependencies. The specified file can not be found..

**Comment by Thomas-Julien-SAHUT** [comment]
UserEntityViewModelTest doesn't work now: issue on the UserEntityViewModelTest.Initialize about Castle.DynamicProxy.InvalidProxyConstructorArgumentsException: Can not instantiate proxy of class: JhipsterXamarin.Services.UserEntityService`1[[JhipsterXamarin.Models.UserModel, JhipsterXamarin, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].

**Comment by Thomas-Julien-SAHUT** [comment]
Unit Test Are good for a review now.

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Utilities/ActionResultExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+namespace JhipsterXamarin.Utilities
+{
+    public static class ActionResultExtensions
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Utilities/ActionResultExtensions.cs]: To be deleted

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Utilities/ActionResultUtil.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+﻿using AutoMapper;
+using Microsoft.AspNetCore.Mvc;
+
+namespace JhipsterXamarin.Utilities
+{
+    public static class ActionResultUtil
+    {
+        public static ActionResult WrapOrNotFoundAsDto<TDtoType>(object value, IMapper mapper)
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Utilities/ActionResultUtil.cs]: To be deleted

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Utilities/ActionResultWithHeaders.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using System;
+using System.Threading.Tasks;
+
+namespace JhipsterXamarin.Utilities
+{
+        public class ActionResultWithHeaders : ActionResult {
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Utilities/ActionResultWithHeaders.cs]: To be deleted

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Utilities/HeaderUtil.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
+﻿using Microsoft.AspNetCore.Http;
+
+namespace JhipsterXamarin.Utilities
+{
+    public static class HeaderUtil
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Utilities/HeaderUtil.cs]: To be deleted

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/ViewModels/UserEntityViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+using JhipsterXamarin.Services;
+using MvvmCross.Commands;
+using MvvmCross.Navigation;
+using MvvmCross.ViewModels;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace JhipsterXamarin.ViewModels
+{
+    public class UserEntityViewModel : MvxViewModel
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/ViewModels/UserEntityViewModel.cs]: why this code is in this PR  ?

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Views/UserEntityView.xaml [hunk 12]
```diff
@@ -0,0 +1,43 @@
+﻿<?xml version="1.0" encoding="UTF-8"?>
+<views:MvxContentPage xmlns:views="clr-namespace:MvvmCross.Forms.Views;assembly=MvvmCross.Forms"
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Views/UserEntityView.xaml]: why this code is in this PR

File: JhipsterXamarin/UnitTestJhispterXamarin/Tests/LoginTest.cs [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+            base.Setup(); // from MvxIoCSupportingTest
+
+            var mockAuthService = new Mock<IAuthenticationService>();
+            var mockNavLoginService = new Mock<IMvxNavigationService>();
+
+            loginViewModel = new LoginViewModel(mockNavLoginService.Object, mockAuthService.Object);
+
+            Ioc.RegisterSingleton<IAuthenticationService>(mockAuthService.Object);
+            Ioc.RegisterSingleton<IMvxNavigationService>(mockNavLoginService.Object);
+        }
```
Comment (nicolas63) [inline JhipsterXamarin/UnitTestJhispterXamarin/Tests/LoginTest.cs]: add a line break

File: JhipsterXamarin/UnitTestJhispterXamarin/Tests/MyEntityViewModelTest.cs [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+            base.Setup(); // from MvxIoCSupportingTest 
+
+            Ioc.RegisterSingleton<IMvxNavigationService>(mockNavService.Object);
+            Ioc.RegisterSingleton<IMyEntityService>(myEntityService.Object);
+
+            myEntityViewModel = new MyEntityViewModel(mockNavService.Object, myEntityService.Object);
+            await myEntityViewModel.Initialize();
+
+            myEntityViewModel.ListElement = myEntities;
+        }
```
Comment (nicolas63) [inline JhipsterXamarin/UnitTestJhispterXamarin/Tests/MyEntityViewModelTest.cs:63]: add a line break

File: JhipsterXamarin/UnitTestJhispterXamarin/Tests/MyEntityViewModelTest.cs [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        {
+            myFirstEntityModelSimple.Name = "the first entity";
+            myFirstEntityModelSimple.Age = 1;
+
+            mySecondEntityModelSimple.Name = "the second entity";
+            mySecondEntityModelSimple.Age = 2;
+
+            MyEntityModelSimple myThirdEntityModelSimple = new MyEntityModel();
+            myThirdEntityModelSimple.Name = "the third entity";
+            myThirdEntityModelSimple.Age = 3;
```
Comment (nicolas63) [inline JhipsterXamarin/UnitTestJhispterXamarin/Tests/MyEntityViewModelTest.cs]: Can you use AutoFixture to generate this entities ?

**Review by nicolas63 — CHANGES_REQUESTED** [review]

File: JhipsterXamarin/UnitTestJhispterXamarin/Tests/UserEntityViewModelTest.cs [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+            userEntityViewModel = new UserEntityViewModel(mockNavLoginService.Object, mockUserEntityService.Object);
+
+            Ioc.RegisterSingleton<IUserEntityService<UserModel>>(mockUserEntityService.Object);
+            Ioc.RegisterSingleton<IMvxNavigationService>(mockNavLoginService.Object);
+        }
+
+        protected override void AdditionalSetup()
+        {
+            admin = new UserModel();
+            admin.Id = "admin";
```
Comment (nicolas63) [inline JhipsterXamarin/UnitTestJhispterXamarin/Tests/UserEntityViewModelTest.cs]: This do not compile. Can you fix it ? We also need to add UT in build step of the github action

**Review by nicolas63 — CHANGES_REQUESTED** [review]

**Comment by nicolas63** [comment]
The actual coverage on the tested view model is close to 60%. we need to get 80%.

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Logs.cs [hunk 17]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using MvvmCross;
+using MvvmCross.Logging;
+
+namespace JhipsterXamarin
+{
+    public static class Logs
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Logs.cs]: To be deleted

File: JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Constants/ErrorConstants.cs [hunk 18]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace JhipsterXamarin.Constants
+{
+    public static class ErrorConstants
```
Comment (nicolas63) [inline JhipsterXamarin/JhipsterXamarin/JhipsterXamarin/Constants/ErrorConstants.cs]: Merge conflict ? This class is already in shared project

**Review by nicolas63 — CHANGES_REQUESTED** [review]

**Review by nicolas63 — APPROVED** [review]

**Pull request merged**